src/OrbitBase/include/OrbitBase/AnyErrorOf.h

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2023 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_ANY_ERROR_OF_H_
+#define ORBIT_BASE_ANY_ERROR_OF_H_
+
+#include <string>
+#include <utility>
+#include <variant>
+
+#include "OrbitBase/ParameterPackTrait.h"
+
+namespace orbit_base {
+
+// A wrapper around `std::variant` holding one instance of multiple possible error types. It's
+// mainly meant to be used with `Result` (`Result<T, AnyErrorOf<E1, E2>`) in cases where a function
+// may return one of multiple possible error types. `AnyErrorOf<E1,E2>` has a `.message()` member
+// function that returns the error message of the holding error by forwarding the call to the
+// `.message()` function of the holding error.
+//
+// `AnyErrorOf` behaves like `std::variant` except for the following properties:
+// 1. `ErrorTypes...` may not have duplicate types - all types must be unique.
+// 2. An empty list of `ErrorTypes...` is not allowed.
+// 3. All error types must have a `.message()` function that's either marked `const` or `static` and
+//    that returns something that's convertible to `std::string`.
+// 4. `AnyErrorOf<T1s...>` can be converted into `AnyErrorOf<T2s...>` if `T2s` contains at least all
+//    the types present in `T1s`. The order of the types has no meaning.
+// 5. `AnyErrorOf` can be directly compared (equality and inequality) to a value of one of its error
+//    types if the given error type defines an equality comparison operator. The compared values are
+//    considered equal if `AnyErrorOf` holds a value of the comparing error type and if the equality
+//    comparison operator returns true.
+template <typename... ErrorTypes>
+class AnyErrorOf : public std::variant<ErrorTypes...> {
+  using Base = std::variant<ErrorTypes...>;
+
+ public:
+  using Base::Base;
+  using Base::operator=;
+
+  static_assert(ParameterPackTrait<AnyErrorOf, ErrorTypes...>::kSize >= 1,
+                "AnyError<> (AnyErrorOf with no error types) is not allowed.");
+
+  static_assert(!ParameterPackTrait<AnyErrorOf, ErrorTypes...>::kHasDuplicates,
+                "AnyError<ErrorTypes...> must not have duplicate error types.");
+
+  template <typename... Types>
+  constexpr static bool CanBeConstructedFromTypesAndIsNotCopy() {
```

Comment:
And that's better how?

Reply:
I have to admit that I'm not a big fan of making this a constant because I was really trying to push for more modern compile time programming with values and functions. But I also don't feel too strongly about this particular case and I want to move this forward, so I changed it to a constant - for the discussion to be continued at another time. :grinning:

Reply:
Yeah, thank you. The only nit: the empty argument round brackets can be omitted in lambda declaration. So it's `[] {} `, not `[](){}`.